### PR TITLE
[5.9] GenericSpecializer: fix a crash with -cross-module-optimization

### PR DESCRIPTION
**Pull request by eeckstein**
If a thunk needs to be generated for a partial_apply specialization, but the specialization doesn't have a body, bail.

rdar://107165121

This is a cherry-pick of https://github.com/apple/swift/pull/64632

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by atrick** [comment]
Why is it bad to create a thunk for an external function? The included test case passes for me (without the corresponding fix on main), so I can't tell what the issue is or whether this fixes it.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 1]
```diff
@@ -3119,6 +3119,12 @@ void swift::trySpecializeApplyOfGeneric(
                             << SpecializedF->getLoweredFunctionType() << "\n");
     NewFunctions.push_back(SpecializedF.getFunction());
   }
+  if (replacePartialApplyWithoutReabstraction &&
+      SpecializedF.getFunction()->isExternalDeclaration()) {
+    // Cannot create a tunk without having the body of the function.
```
Comment (atrick) [inline lib/SILOptimizer/Utils/Generics.cpp:3124]: If you create a main PR to fix the "tunk" typo, you could also explain that thunk creation needs to access the callee function's SIL arguments.

**Review by atrick — APPROVED** [review]
Looks good for 5.9

**Review by tbkka — APPROVED** [review]

**Pull request merged**